ext/standard/proc_open.c

```diff
@@ -1203,38 +1248,46 @@ PHP_FUNCTION(proc_open)
 
 		close_descriptor(descriptors[i].childend);
 
-		if (descriptors[i].is_pipe) {
-			switch (descriptors[i].mode_flags) {
+		if (descriptors[i].type != DESCRIPTOR_TYPE_STD) {
+			if (descriptors[i].type == DESCRIPTOR_TYPE_PIPE) {
```

Comment:
You should be able to move this code into the else branch below and reduce the scope of mode_string.